util/EIP-152/src/lib.rs

```diff
@@ -38,58 +43,43 @@ const IV: [u64; 8] = [
 	0x510e527fade682d1, 0x9b05688c2b3e6c1f, 0x1f83d9abfb41bd6b, 0x5be0cd19137e2179,
 ];
 
-
-#[inline(always)]
-/// The G mixing function. See https://tools.ietf.org/html/rfc7693#section-3.1
-fn g(v: &mut [u64], a: usize, b: usize, c: usize, d: usize, x: u64, y: u64) {
-	v[a] = v[a].wrapping_add(v[b]).wrapping_add(x);
-	v[d] = (v[d] ^ v[a]).rotate_right(32);
-	v[c] = v[c].wrapping_add(v[d]);
-	v[b] = (v[b] ^ v[c]).rotate_right(24);
-	v[a] = v[a].wrapping_add(v[b]).wrapping_add(y);
-	v[d] = (v[d] ^ v[a]).rotate_right(16);
-	v[c] = v[c].wrapping_add(v[d]);
-	v[b] = (v[b] ^ v[c]).rotate_right(63);
-}
-
-/// The Blake2 compression function F. See https://tools.ietf.org/html/rfc7693#section-3.2
-/// Takes as an argument the state vector `h`, message block vector `m`, offset counter `t`, final
-/// block indicator flag `f`, and number of rounds `rounds`. The state vector provided as the first
-/// parameter is modified by the function.
-pub fn compress(h: &mut [u64; 8], m: [u64; 16], t: [u64; 2], f: bool, rounds: usize) {
-	let mut v = [0u64; 16];
-	v[..h.len()].copy_from_slice(h);    // First half from state.
-	v[h.len()..].copy_from_slice(&IV);  // Second half from IV.
-
-	v[12] ^= t[0];
-	v[13] ^= t[1];
-
-	if f {
-		v[14] = !v[14] // Invert all bits if the last-block-flag is set.
-	}
-	for i in 0..rounds {
-		// Message word selection permutation for this round.
-		let s = &SIGMA[i % 10];
-		g(&mut v, 0, 4, 8, 12, m[s[0]], m[s[1]]);
-		g(&mut v, 1, 5, 9, 13, m[s[2]], m[s[3]]);
-		g(&mut v, 2, 6, 10, 14, m[s[4]], m[s[5]]);
-		g(&mut v, 3, 7, 11, 15, m[s[6]], m[s[7]]);
-
-		g(&mut v, 0, 5, 10, 15, m[s[8]], m[s[9]]);
-		g(&mut v, 1, 6, 11, 12, m[s[10]], m[s[11]]);
-		g(&mut v, 2, 7, 8, 13, m[s[12]], m[s[13]]);
-		g(&mut v, 3, 4, 9, 14, m[s[14]], m[s[15]]);
+// raw fn pointer type, can be cast to any fn type
+type FnRaw = *mut ();
+// Blake2bF fn pointer type
+type Blake2bFn = fn(&mut [u64; 8], [u64; 16], [u64; 2], bool, usize);
+
+// holds the pointer to the compress function
+static FN: AtomicPtr<()> = AtomicPtr::new(detect as FnRaw);
```

Comment:
Did the `ifunk` trick actually make a difference in benchmarks?

Reply:
It's hard to say. Sometimes it's better sometimes it's worse.

Reply:
If there is no clear benefit, let's remove it then to avoid maintenance burden.
Also please fix the funny indentation (see github diff).